### PR TITLE
lib/modules: Add function to create option alias that respects priority

### PR DESCRIPTION
**Pull request by cdepillabout**
###### Motivation for this change

This commit adds a function `mkAliasOptionModuleWithPriority`.  This function will make an alias to an existing option and copy over the priority.  (This is in contrast to `mkAliasOptionModule` which does not copy over the priority.)

This functionality is needed for PRs like #53041 (specifically https://github.com/NixOS/nixpkgs/pull/53041#commitcomment-31825338).

Here's the use-case:

There exists an option `powerManagement.cpuFreqGovernor`.  On some systems, this is set in `hardware-configuration.nix` by `nixos-generate-config`:

```nix
{
  powerManagement.cpuFreqGovernor = mkDefault "powersave";
}
```

Note how it gets a low priority.  The end-user (or other modules) should be able to easily override it.

In #53041, I wanted to add options to set the CPU max and min frequency:

```nix
{
  powerManagement.cpufreq.max = 2200000;
  powerManagement.cpufreq.min = 1500000;
}
```

I thought it was silly to have the options called both `powerManagement.cpuFreqGovernor` and `powerManagement.cpufreq.max`, so I renamed `powerManagement.cpuFreqGovernor` to `powerManagement.cpufreq.governor`.  This is fine, but it will break for people that are using `hardware-configuration.nix` generated with an older version of `nixos-generate-config` (since in their config it is called `powerManagement.cpuFreqGovernor`).

In order to accommodate those people, I added the alias `powerManagement.cpuFreqGovernor` to `powerManagement.cpufreq.governor`:

```nix
mkAliasOptionModule [ "powerManagement" "cpuFreqGovernor" ] [ "powerManagement" "cpufreq" "governor" ]
```

This should let people use their older `hardware-configuration.nix` with the `powerManagement.cpuFreqGovernor` option.

However, there is one edge case this doesn't work for.

Imagine that the user has `powerManagement.cpuFreqGovernor` set in `hardware-configuration.nix`:

```nix
{
  powerManagement.cpuFreqGovernor = mkDefault "powersave";
}
```

The user then tries to set `powerManagement.cpufreq.governor` to something else in `configuration.nix`:

```nix
{
  powerManagement.cpufreq.governor = "ondemand";
}
```

This actually will fail.  The user will get an error that `powerManagement.cpufreq.governor` is defined two places (which is true, but the definition in `hardware-configuration.nix` should have low priority).

What's going on here?  Well, the `mkAliasOptionModule` function doesn't actually copy over the priority of the alias option to the normal option.  So to the nix module system, it really does look like `powerManagement.cpufreq.governor` is being set two different places.

This PR adds a function `mkAliasOptionModuleWithPriority` that does correctly(?) copy over the priority from the alias option to the normal option.

Tests have also been added to make sure this is working correctly.

Pinging @Infinisil since he helped a lot with #53041.  Also pinging @edolstra @nbp since they seem to be code-owners of the stuff in `lib/`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by nbp — CHANGES_REQUESTED** [review]
Thanks! For some reasons I thought this was the case but this might be a really long time ago...
A minor change and making the test cases more robust and this should be good for landing.

File: lib/modules.nix [hunk 1]
```diff
@@ -556,8 +556,21 @@ rec {
...
-    mkIf (isOption option && option.isDefined) (wrap (mkMerge option.definitions));
+    mkAliasIfDef option (wrap (mkMerge option.definitions));
 
+  # Similar to mkAliasAndWrapDefinitions but copies over the priority from the
+  # option as well.
+  #
+  # If a priority is not set, it assumes a priority of 100.
+  mkAliasAndWrapDefsWithPriority = wrap: option:
+    let
+      defaultPrio = 100;
```
[the remark this answers is not in the document]
Comment (cdepillabout) [reply lib/modules.nix]: Good idea.  I've done this in commit f24e2d0721c.

File: lib/modules.nix [hunk 2]
```diff
@@ -690,7 +703,16 @@ rec {
...
+  /* Like ‘mkAliasOptionModule’, but copy over the priority of the option as well. */
+  mkAliasOptionModuleWithPriority = from: to: doRename {
+    inherit from to;
+    visible = true;
+    warn = false;
+    use = id;
+    withPriority = true;
+  };
+
+  doRename = { from, to, visible, warn, use, withPriority ? false }:
```
[the remark this answers is not in the document]
Comment (cdepillabout) [reply lib/modules.nix]: Will do.

File: lib/tests/modules.sh [hunk 3]
```diff
@@ -149,6 +149,9 @@ checkConfigOutput "1 2 3 4 5 6 7 8 9 10" config.result ./loaOf-with-long-list.ni
 # Check loaOf with many merges of lists.
 checkConfigOutput "1 2 3 4 5 6 7 8 9 10" config.result ./loaOf-with-many-list-merges.nix
 
+# Check mkAliasOptionModuleWithPriority.
+checkConfigOutput "true" config.enable ./alias-with-priority.nix
```
[the remark this answers is not in the document]
Comment (cdepillabout) [reply lib/tests/modules.sh]: This is a good idea.  I added a test that shows that the alias option is actually able to override the non-alias option as long as the priority is high enough.

**Comment by cdepillabout** [comment]
@nbp Thanks for the review and the good suggestions.  I have fixed the things you've pointed out.

This PR should be ready for a final review and hopefully merging.

**Review by nbp — APPROVED** [review]
Sorry for the delay and thanks for the patches. These are great changes!

**Pull request merged**